sys/cheri/cheri_usercap.c

```diff
@@ -96,13 +102,57 @@ _cheri_capability_build_user_rwx(uint32_t perms, ptraddr_t basep, size_t length,
     off_t off, const char* func __unused, int line __unused, bool exact)
 {
 	void * __capability tmpcap;
+#ifdef INVARIANTS
+	vm_map_entry_t entry;
+	vm_map_t map;
+	vm_offset_t reservation;
+
+	if (SV_CURPROC_FLAG(SV_CHERI)) {
+		map = &curproc->p_vmspace->vm_map;
+		vm_map_lock_read(map);
+		KASSERT(vm_map_lookup_entry(map, basep, &entry),
+		    ("%s:%d: vm_map does not contain basep 0x%zx "
+		    "(length 0x%zu, offset 0x%ju)", func, line,
+		    (size_t)basep, length, (uintmax_t)off));
+		reservation = entry->reservation;
+		for( ; basep + length > entry->end;
+		    entry = vm_map_entry_succ(entry)) {
+			/*
+			 * Check that the created capability is within a
+			 * single reservation.  This ensures we don't
+			 * make capabilities that might alias with a
+			 * later mapping.
+			 */
+			KASSERT((map->flags & MAP_RESERVATIONS) == 0 ||
+			    entry->reservation == reservation,
+			    ("Can't create a capability that spans reservations"));
 
-	tmpcap = cheri_setoffset(cheri_andperm(cheri_setbounds(
-	    cheri_setoffset(userspace_root_cap, basep), length), perms), off);
+			/*
+			 * XXX: Disallow quarantined or abandoned pages?
+			 *
+			 * XXX: Require page permissions to be a superset of
```

Comment:
You would want to validate against maxprot in the VM map entry?

Reply:
I've updated the comment to reflect this.